iter13

repriest opened the pull request:
- changed id to uuid
- return existing shortURL on conflict in pg
- utils for handlers

<!-- GitButler Footer Boundary Top -->
---
This is **part 13 of 15 in a stack** made with GitButler:
- <kbd>&nbsp;15&nbsp;</kbd> #15 
- <kbd>&nbsp;14&nbsp;</kbd> #14 
- <kbd>&nbsp;13&nbsp;</kbd> #13 👈 
- <kbd>&nbsp;12&nbsp;</kbd> #12 
- <kbd>&nbsp;11&nbsp;</kbd> #11 
- <kbd>&nbsp;10&nbsp;</kbd> #10 
- <kbd>&nbsp;9&nbsp;</kbd> #9 
- <kbd>&nbsp;8&nbsp;</kbd> #8 
- <kbd>&nbsp;7&nbsp;</kbd> #7 
- <kbd>&nbsp;6&nbsp;</kbd> #6 
- <kbd>&nbsp;5&nbsp;</kbd> #5 
- <kbd>&nbsp;4&nbsp;</kbd> #4 
- <kbd>&nbsp;3&nbsp;</kbd> #3 
- <kbd>&nbsp;2&nbsp;</kbd> #2 
- <kbd>&nbsp;1&nbsp;</kbd> #1 
<!-- GitButler Footer Boundary Bottom -->

Commit:
check db with /ping handler

- db ping context timeout 5sec

Commit:
validate empty dsn

Commit:
minor fixes

Commit:
add DB store

- added pg / file / memory stores
- added Close to storage interface
- use pg -> file -> memory to store
- updated tests

Commit:
fix tests

- init flags in TestMain
- check rows.Err

Commit:
add ShortenBatchHandler

- add batch append to storage

Commit:
fix tests

Commit:
fix tests

Commit:
log body

Commit:
more logs

Commit:
fix id

- accept uuid

Commit:
return shorturl on conflict

- changed id to uuid
- return existing shortURL on conflict in pg
- utils for handlers

Commit:
fix tests

Commit:
fix test

Commit:
update logic

- custom err struct with existing url
- fix storage append to use error

Commit:
fix content json

Commit:
update err conflict logic

vokinneberg commented on cmd/shortener/main.go:
```diff
@@ -27,11 +30,33 @@ func initLogger(cfg *config.Config) error {
+		st = storage.NewMemoryStorage()
+	}
+
+	return storage.NewRepository(st)
```
Мне кажется стоит единообразно подойти к возврату значения. Либо наоборот, инициализировать в каждой ветке `st`  и в конце делать `return  storage.NewRepository(st)`.
```suggestion
	if cfg.DatabaseDSN != "" {
		st, err := storage.NewPostgresStorage(cfg.DatabaseDSN)
		if err != nil {
			return nil, fmt.Errorf("could not connect to postgres: %w", err)
		}
		return storage.NewRepository(st)

	} else if cfg.FileStoragePath != "" {
		st, err := storage.NewFileStorage(cfg.FileStoragePath)
		if err != nil {
			return nil, fmt.Errorf("could not open file %s: %w", cfg.FileStoragePath, err)
		}
		return storage.NewRepository(st)

	} else {
		return storage.NewRepository(storage.NewMemoryStorage())
	}
```

vokinneberg commented on cmd/shortener/main.go:
```diff
@@ -40,6 +65,8 @@ func initRouter(cfg *config.Config, st *storage.Repository) *chi.Mux {
 	r.Post("/", logger.RequestLogger(zipper.GzipMiddleware(h.ShortenHandler)))
 	r.Get("/{id}", logger.ResponseLogger(zipper.GzipMiddleware(h.ExpandHandler)))
 	r.Post("/api/shorten", logger.RequestLogger(zipper.GzipMiddleware(h.ShortenJSONHandler)))
+	r.Get("/ping", logger.RequestLogger(zipper.GzipMiddleware(h.PingHandler)))
```
Нужен ли нам в этом случае Gzip middleware?

vokinneberg replied:
Так же здесь можно добавить создать группу хэнедлеров и добавить middleware один раз:
```go
	r.Get("/ping", h.PingHandler)
	r.Group(func(r chi.Router) {
		r.Use(logger.RequestLogger)
		r.Use(zipper.GzipMiddleware)
		r.Post("/", h.ShortenHandler)
		r.Get("/{id}", h.ExpandHandler)
		r.Post("/api/shorten", h.ShortenJSONHandler)
		r.Post("/api/shorten/batch", h.ShortenBatchHandler)
	})
```

Правда для этого необходимо будет поменять сигнатуру middleware:
```go
func RequestLogger(h http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()

		var bodyBytes []byte
		if r.Body != nil {
			bodyBytes, _ = io.ReadAll(r.Body)
			r.Body = io.NopCloser(bytes.NewBuffer(bodyBytes))
		}

		Log.Info("incoming HTTP request",
			zap.String("method", r.Method),
			zap.String("uri", r.RequestURI),
			zap.String("content_type", r.Header.Get("Content-Type")),
			zap.ByteString("body", bodyBytes),
		)

		h.ServeHTTP(w, r)

		Log.Info("request completed",
			zap.Duration("duration", time.Since(start)),
			zap.String("content_type", r.Header.Get("Content-Type")),
			zap.String("location", r.Header.Get("Location")),
		)
	})
}
```

vokinneberg commented on cmd/shortener/main_test.go:
```diff
@@ -96,14 +100,14 @@ func TestExpandHandler(t *testing.T) {
 	}{
 		{
 			name:       "Valid URL",
```
Я не рекомендую использовать пробелы в названиях сабтестов или табличных тестов, так как Go автоматически заменяет пробелы на `_` в выводе, это затрудняет поиск по названию теста. Стоит избегать пробелов или сразу заменять их на `_`.

vokinneberg commented on internal/config/config.go:
```diff
@@ -115,3 +121,21 @@ func validateFileStoragePath(path string) error {
 	return nil
 }
+
+func validateDatabaseDSN(dsn string) error {
```
Я бы все-таки делал это в момент подключения к БД в `main(...)` - это было бы более наглядно и можно было бы использовать существующее подключение. В противном случае, мы создадим новое подключение и не будет гарантии, что оно рабочее.

vokinneberg commented on internal/handlers/handlers.go:
```diff
@@ -57,18 +30,31 @@ func (h *Handler) ShortenHandler(w http.ResponseWriter, r *http.Request) {
+	// check existing shortURL
 	err = h.st.AddNewEntry(shortURL, longURL)
 	if err != nil {
+		var urlConflictError *t.URLConflictError
+		if errors.As(err, &urlConflictError) { // get instance of URLConflictError if err matches
+			if urlConflictError.ShortURL != "" {
+				responseURL = h.cfg.BaseURL + "/" + urlConflictError.ShortURL
+			} else {
+				responseURL = h.cfg.BaseURL + "/" + shortURL
+			}
+			w.WriteHeader(http.StatusConflict)
+		} else {
+			http.Error(w, "Could not write URL to storage", http.StatusInternalServerError)
+			return
+		}
+	} else {
+		w.WriteHeader(http.StatusCreated)
+	}
```
Мне кажется отсутствие `else` там, где его можно опустить улучшает читаемость кода.
```suggestion
		var urlConflictError *t.URLConflictError
		if errors.As(err, &urlConflictError) { // get instance of URLConflictError if err matches
			if urlConflictError.ShortURL != "" {
				responseURL = h.cfg.BaseURL + "/" + urlConflictError.ShortURL
			} else {
				responseURL = h.cfg.BaseURL + "/" + shortURL
			}
			w.WriteHeader(http.StatusConflict)
		}
		http.Error(w, "Could not write URL to storage", http.StatusInternalServerError)
		return
	}
	w.WriteHeader(http.StatusCreated)
```

repriest replied:
мне кажется, w.WriteHeader(http.StatusCreated) перезапишет w.WriteHeader(http.StatusConflict)
но суть понял

vokinneberg commented on cmd/shortener/main.go:
```diff
@@ -27,11 +30,33 @@ func initLogger(cfg *config.Config) error {
-	st, err := storage.NewRepository(cfg.FileStoragePath)
-	if err != nil {
-		return nil, err
+	var st t.Storage
```
Было бы здорово реализовать зависимость от Storage через interface:
```go
type Storage interface {
	AddNewEntry(shortURL string, originalURL string) error
	BatchAppend(entries []t.URLEntry) error
	Close() error
}
```

vokinneberg commented on internal/storage/file/file.go:
```diff
@@ -0,0 +1,87 @@
+package file
```
Не понимаю какое назначение пакета file и т.д. в пакете `storage`? Мне кажется если избавиться от всех  вложенных пакетов будет только проще. Все типы Storage разные и не конфликтуют друг с другом.

vokinneberg commented on internal/handlers/types.go:
```diff
@@ -0,0 +1,33 @@
+	st  *storage.Repository
+}
+
+func NewHandler(cfg *config.Config, st *storage.Repository) *Handler {
```
Было бы здорово реализовать зависимость от Storage через interface:
```go
type Storage interface {
	Load() ([]URLEntry, error)
	Append(entry URLEntry) error
	BatchAppend(entries []URLEntry) error
	Close() error
}
```
```suggestion
func NewHandler(cfg *config.Config, st *Storage) *Handler {
```

vokinneberg commented on internal/storage/postgres/postgres.go:
```diff
@@ -0,0 +1,131 @@
+	db *sql.DB
+}
+
+func NewPgStorage(dsn string) (t.Storage, error) {
```
Здесь нет необходимости возвращать тип интерфейса. В идеале реализация вообще не обязана знать об этом интерфейсе, так как если тип реализует методы интерфейса он автоматически реализует сам интерфейс. При этом он будет реализовать не только интерфейс, который объявляет все методы, оно и какой-то один метод, Например:
```go
type BatchAppender interface {
	BatchAppend(entries []URLEntry) error
}
```

Возвращая интерфейс ты искусственно ограничиваешь клиентский код.
 
```suggestion
type PGStorage struct {
	db *sql.DB
}

func NewPgStorage(dsn string) (*PGStorage, error) {
```

vokinneberg commented on internal/storage/postgres/postgres.go:
```diff
@@ -0,0 +1,131 @@
+		return nil, fmt.Errorf("failed to open database: %w", err)
+	}
+
+	_, err = db.Exec(`
```
Сейчас это ОК, но в будущем можно будет подумать о миграциях БД.

vokinneberg commented on internal/storage/types/types.go:
```diff
@@ -0,0 +1,29 @@
+	"fmt"
+)
+
+var ErrURLExists = errors.New("URL already exists")
```
Предложение по поводу плоской структуры касается и этого фала. И не нужен будет алиас `t`.

vokinneberg commented on cmd/shortener/main_test.go:
```diff
@@ -13,18 +13,26 @@ import (
-		FileStoragePath: "url_store.json",
+var cfg *config.Config
+
+func TestMain(m *testing.M) {
```
Возможно что-то упустил, но по-моему тут не хватает покрытия новых кейсов?

vokinneberg reviewed (changes requested):
Первое ревью третьего спринта! 🎉 В этом спринте наше приложение стало ещё сложнее, добавился новый слой БД 💪 Хотелось бы добавить несколько комментариев относительно твоего кода и сделать несколько предложений по улучшению. 🤓

**Сильные стороны твоего PR:**

- Код функционален и корректен! 🚀
- Ты реализовал обратную совместимость хранилищ. 👍

**На что я бы рекомендовал обратить внимание и что можно улучшить:**

- Можно создать группу хэндлеров и добавить middleware один раз.
- Я не рекомендую использовать пробелы в названиях сабтестов или табличных тестов, так как Go автоматически заменяет пробелы на `_` в выводе, это затрудняет поиск по названию теста. Стоит избегать пробелов или сразу заменять их на `_`.
- Я бы проверял `Ping()` к БД в момент подключения к БД в `main(...)` — это было бы более наглядно и можно было бы использовать существующее подключение. В противном случае, мы создадим новое подключение и не будет гарантии, что оно рабочее.
- Мне кажется, отсутствие `else` там, где его можно опустить, улучшает читаемость кода.
- В хендлерах было бы здорово реализовать зависимость от Storage через interface.
- Не понимаю, какое назначение пакета file и т.д. в пакете `storage`? Мне кажется, если избавиться от всех вложенных пакетов, будет только проще. Все типы Storage разные и не конфликтуют друг с другом.
- В конструкторе Storage нет необходимости возвращать тип интерфейса. В идеале реализация вообще не обязана знать об этом интерфейсе, так как если тип реализует методы интерфейса, он автоматически реализует сам интерфейс. При этом он будет реализовывать не только интерфейс, который объявляет все методы, но и какой-то один метод.
- Сейчас это ОК, но в будущем можно будет подумать о миграциях БД.
- Возможно что-то упустил, но по-моему тут не хватает тестового покрытия новых кейсов?

Commit:
review fixes

1. chi group и use для middleware
2. batch handler test
3. omit Repository, use Storage iface
4. validate db ping on pg init

Commit:
fix test

vokinneberg reviewed (approved):
Поздравляю с успешным ревью третьего спринта! 🎉
